src/query/executor.rs

```diff
@@ -0,0 +1,162 @@
+use crate::ff::{Field, FieldType, Fp31};
+use crate::helpers::messaging::Gateway;
+use crate::helpers::negotiate_prss;
+use crate::helpers::query::{QueryConfig, QueryType};
+use crate::protocol::context::SemiHonestContext;
+use crate::protocol::Step;
+use crate::secret_sharing::Replicated;
+use crate::task::JoinHandle;
+use futures::Stream;
+use rand::rngs::StdRng;
+use rand_core::SeedableRng;
+#[cfg(all(feature = "shuttle", test))]
+use shuttle::future as tokio;
+
+pub trait Result: Send {
+    fn into_bytes(self: Box<Self>) -> Vec<u8>;
+}
+
+impl<F: Field> Result for Vec<Replicated<F>> {
+    fn into_bytes(self: Box<Self>) -> Vec<u8> {
+        // todo Result
+        let mut r = Vec::with_capacity(2 * self.len() * F::SIZE_IN_BYTES as usize);
+        for share in self.into_iter() {
+            let mut slice = [0u8; 32]; // we don't support fields > 16 bytes
+            let written = share.serialize(&mut slice).unwrap();
+            r.extend_from_slice(&slice[..written]);
+        }
+
+        r
+    }
+}
+
+#[cfg(any(test, feature = "test-fixture"))]
+async fn test_multiply<F: Field, S: Stream<Item = Vec<u8>> + Send + Unpin>(
+    ctx: SemiHonestContext<'_, F>,
+    mut input: S,
+) -> Vec<Replicated<F>> {
+    use crate::protocol::basics::SecureMul;
+    use crate::protocol::RecordId;
+    use futures_util::StreamExt;
+
+    let mut results = Vec::new();
+    while let Some(v) = input.next().await {
+        // convert bytes to replicated shares
+        let shares = v.chunks(2 * F::SIZE_IN_BYTES as usize).map(|chunk| {
+            // TODO fix with replicated serialization
+            let left = F::deserialize(&chunk[..=F::SIZE_IN_BYTES as usize]).unwrap();
+            let right = F::deserialize(&chunk[F::SIZE_IN_BYTES as usize..]).unwrap();
+
+            Replicated::new(left, right)
+        });
+
+        // multiply pairs
+        let mut a = None;
+        let mut record_id = 0_u32;
+        for share in shares {
```

Comment:
might be more obvious flow if something like
```rust
for (record_id, shares) in shares.chunks(2).enumerate() {
    assert_eq!(shares.len(), 2);
    let share_a = shares[0];
    let share_b = shares[1];

Reply:
I don't think I can make it work because `chunks` need a container to store data and it is not implemented for iterators, only for slices.

Reply:
ah...yeah i understand. i was expecting it to be able to coerce into `Vec`